src/jit/morph.cpp

```diff
@@ -7740,261 +7675,80 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
  *
  *  Transform the given GT_CALL tree for tail call code generation.
  */
-void Compiler::fgMorphTailCallViaHelper(GenTreeCall* call, void* pfnCopyArgs)
+void Compiler::fgMorphTailCallViaHelper(GenTreeCall* call, CORINFO_TAILCALL_HELP& help)
 {
-#if defined(_TARGET_UNIX_)
-    noway_assert(!"Slow tail calls not supported on non-Windows platforms.");
-#endif
-
     JITDUMP("fgMorphTailCallViaHelper (before):\n");
     DISPTREE(call);
 
-    // The runtime requires that we perform a null check on the `this` argument before
-    // tail calling  to a virtual dispatch stub. This requirement is a consequence of limitations
-    // in the runtime's ability to map an AV to a NullReferenceException if
-    // the AV occurs in a dispatch stub that has unmanaged caller.
-    if (call->IsVirtualStub())
-    {
-        call->gtFlags |= GTF_CALL_NULLCHECK;
-    }
+    // TODO: Some comments (check old document and see what still applies)
```

Comment:
I think that it would be really useful here (or, rather, up in the function header) to summarize the transformation that's being made. It took me multiple reads of the design document, as well as the code here, to begin to understand. At least from a JIT point of view, a summary like this would be useful (corrected, of course, where I may still misunderstand):
```
// This transforms the following:
//    {RETURN | ASG}
//        CALL
//            {callTarget}
//            {this}
//            {args}
// to
//    CALL storeArgsThunk ; not tail call
//        callTarget
//        {this}
//        {args}
//    CALL callTarget ; tail call